vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -141,14 +129,65 @@ def _transformed_inputs(self) -> Union[list[State], dict[str, Union[State, Contr
         # have already been classified as legacy in Action._legacy. In future when vm.Action.inputs is deprecated
         # then this will be used for vm.Action instances also.
         TypeAdapter(dict[str, _DotSeparatedStr]).validate_python(self._runtime_args)
-
+        # Do exactly same lookup but we only have "__default__" defined in inputs dictionary so will only ever use that
+        # case
         # User specified arguments runtime_args take precedence over built in reserved arguments. No static arguments
         # ar relevant here, just Dash States. Static arguments values are stored in the state of the relevant
         # _AbstractAction instance.
-        runtime_args = {arg_name: State(*arg_value.split(".")) for arg_name, arg_value in self._runtime_args.items()}
+        # Qn: should this work for legacy actions too? Think about docs.
+        runtime_args = {
+            arg_name: State(*self._transform(arg_value, type="input").split("."))
+            for arg_name, arg_value in self._runtime_args.items()
+        }
 
         return builtin_args | runtime_args
 
+    @staticmethod
+    def _transform(reference: str, type: Literal["output", "input"]) -> str:
+        """Transform a component reference into its mapped property value.
+
+        Handles two formats:
+        - "component-id.component-property" (e.g. "graph-1.figure"): Returns mapped value if exists, else original
+        - "component-id" (e.g. "graph-1"): Returns default mapping from _outputs/_inputs, raises if no default
+
+        Args:
+            reference: Component ID or "component-id.component-property" string
+            type: "input" or "output" to check _inputs or _outputs
+
+        Returns:
+            Mapped property value or original reference for explicit format
+
+        Raises:
+            KeyError: If component not found or no default mapping exists
+            ValueError: If reference format is invalid e.g. "model-id.prop.prop"
+        """
+        property_name = "_outputs" if type == "output" else "_inputs"
+
+        if "." in reference:
+            # LQ: Check whether we need TypeAdapter here or something else. Currently it doesn't catch cases such as
+            # `component-id.prop.prop` well. However, adding below leads to a bunch of unit tests to fail. Check later.
+            # TypeAdapter(Union[list[_DotSeparatedStr], dict[str, _DotSeparatedStr]]).validate_python(reference)
+            component_id, component_property = reference.split(".")
+            if component_id in model_manager and hasattr(model_manager[component_id], property_name):
+                if component_property in getattr(model_manager[component_id], property_name):
+                    return getattr(model_manager[component_id], property_name)[component_property]
+            return reference
```

Comment:
EAFP would make this clearer I think:

```suggestion
            try:
                return getattr(model_manager[component_id], property_name)[component_property]
            except KeyError, AttributeError:
                 # Some comment here explaining what this clause is for - could happen if component_id not in MM
                 # or intputs/outputs not defined or component_property not in that dictionary
                return dependency
```

Reply:
I've kept the try/except structure for both cases (dot-notation and model id only ), but had to add more logic inside, so do take a look if it's easier to read now 👍

Reply:
I like the way you changed it 👍